### PR TITLE
Refactor functions in cloudant

### PR DESCRIPTION
**Pull request by ssh24**
connect to https://github.com/strongloop-internal/scrum-apex/issues/204

### CODE REVIEW

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

**Comment by ssh24** [comment]
@slnode test please

File: lib/cloudant.js [hunk 1]
```diff
@@ -203,7 +189,9 @@ Cloudant.prototype.create = function(model, data, options, cb) {
  *
  * @param {String} model The model name
  * @param {Object} data The model data
+ * @param {Object} options The options object
  * @param {Function} [callback] The callback function
+ * @returns {Function} [create] model create function
```
Comment (jannyHou) [inline lib/cloudant.js]: Does it have a return function? I think it just calls `_insert`

Comment (ssh24) [inline lib/cloudant.js]: It returns the inserted data. Should we not have that as a return @jannyHou ?

Comment (jannyHou) [inline lib/cloudant.js]: I would probably put what get returned in `_insert` function then. Just a nitpick, feel free to do it in your way :p

File: lib/cloudant.js [hunk 2]
```diff
@@ -88,15 +88,13 @@ Cloudant.prototype.connect = function(cb) {
   self.CloudantDriver(self.options, function(err, nano) {
     if (err) return cb(err);
     self.cloudant = nano;
-    if (self.options.database) {
+    if (self.options.database)
       // check if database exists
       self.cloudant.db.get(self.options.database, function(err) {
         if (err) return cb(err);
         return cb(err, self.cloudant);
       });
```
Comment (0candy) [inline lib/cloudant.js]: Could we leave the braces for this if block?

File: intl/en/messages.json [hunk 3]
```diff
@@ -1,4 +1,6 @@
 {
-  "f93ce20d178e2d16f45ed0a1a117b990": "example complete",
-  "5bdcd0e36db5b1100819f9aa25e9bdf0": "Cloudant {{regex}} syntax does not support global"
+  "3cde8cc9bca22c67278b202ab0720106": "No instance with id {0} found for {1}",
+  "5bdcd0e36db5b1100819f9aa25e9bdf0": "Cloudant {{regex}} syntax does not support global",
+  "934c33143a59fd286ff6f7d0b8fb6875": "Invalid settings: \"url\" OR \"username\" AND \"password\" required",
+  "f93ce20d178e2d16f45ed0a1a117b990": "example complete"
```
Comment (b-admike) [inline intl/en/messages.json]: `example complete` doesn't sound like an error; is it used anywhere?

File: lib/cloudant.js [hunk 4]
```diff
@@ -284,269 +271,50 @@ Cloudant.prototype.all = function all(model, filter, options, cb) {
...
-      cb(null, rst);
-    }
-  };
-};
 
+/**
+ * Build query selector
+ *
+ * @param {String} model The model name
+ * @param {Object} mo The model object
```
Comment (jannyHou) [inline lib/cloudant.js]: nitpick: 
The model object generated by `selectModel()`

File: lib/cloudant.js [hunk 5]
```diff
@@ -770,19 +515,19 @@ Cloudant.prototype.updateOrCreate = function(model, data, cb) {
  * @param {String} model The model name
  * @param {Object} where The search criteria
  * @param {Object} data The property/value pairs to be updated
- * @callback {Function} cb Cb function
+ * @param {Object} options The options Object
+ * @param {Function} [cb] The callback function
```
Comment (jannyHou) [inline lib/cloudant.js]: Hmm, seems the right syntax is `@callback {Function} callback` according to current juggler code.

File: lib/cloudant.js [hunk 6]
```diff
@@ -1096,5 +845,321 @@ Cloudant.prototype.updateIndex = function(mo, modelName, cb) {
...
+      propIsObjWithTypeArr = true;
+      return true;
+    }
+    return false;
+  }
+  return newFields.join('.');
+}
+
+/**
+ *  Recursively find function
```
Comment (b-admike) [inline lib/cloudant.js]: Can you re-word this to be more clear?

File: lib/cloudant.js [hunk 7]
```diff
@@ -1096,5 +845,321 @@ Cloudant.prototype.updateIndex = function(mo, modelName, cb) {
...
+    default:
+      newQuery = {};
+      newQuery['$' + op] = cond;
+  }
+  return [newQuery, containsRegex];
+}
+
+/** Build an array of filter
+ *
+ * REVIEW: @param {String} k ?
```
Comment (b-admike) [inline lib/cloudant.js]: I would also like to know what k is!

File: intl/en/messages.json (hunk 3 above)
Comment (b-admike) [inline intl/en/messages.json]: Disregard this. It was there before.

Comment (ssh24) [inline intl/en/messages.json]: @b-admike This is where `example complete` is used. Does not necessarily need to be an error. https://github.com/strongloop/loopback-connector-cloudant/blob/2ba7d7f77a96203c617699c4a5c677f3880689ec/example/example.js#L39

**Pull request merged**